### PR TITLE
Raft providers

### PR DESCRIPTION
**Pull request by kvahed**
# Pull Request Guidelines

Pull requests are an essential collaborative tool for modern software development. 

The below list is intended to help you figure out whether your code is ready to be reviewed
and merged into ArangoDB. The overarching goal is to:

- Reduce the amount of *recurring* defects
- Reduce the impact to the other developer’s time and energy spent on defects in your code
- Increase the overall autonomy and productivity of individual developers

## Acceptance Checklist

*The below list is not exhaustive, think thoroughly whether the provided information is sufficient.*
*Remove options that do not apply*

### Scope & Purpose

*(Can you describe what functional change your PR is trying to effect?)*

- [ ] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [ ] Bug-Fix for a *released version* (did you remember to port this to all relevant release branches?)
- [ ] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [ ] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [ ] The behavior change can be verified via automatic tests

#### Related Information

*(Please reference tickets / specification etc )*

- [ ] There is a GitHub Issue reported by a Community User: 
- [ ] There is an internal planning ticket: 
- [ ] There is a *JIRA Ticket number* (In case a customer was affected / involved): 
- [ ] There is a design document: 

### Testing & Verification

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This PR adds tests that were used to verify all changes:

- [ ] Added **Regression Tests** (Only for bug-fixes) 
- [ ] Added new C++ **Unit Tests** (Either GoogleTest or Catch-Test)
   - Did you add tests for a new *RestHandler* subclass ?
   - Did you add new mocks of underlying code layers to be able to verify your functionality ? 
   - ...
- [ ] Added new **integration tests** (i.e. in shell_server / shell_server_aql)
- [ ] Added new **resilience tests** (only if the feature is impacted by failovers)

Additionally:

- [ ] There are tests in an external testing repository (i.e. node-resilience tests, chaos tests)
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

*(Include link to Jenkins run etc)*

> Think about whether the new code you added is modular enough to be
> easily testable by unit tests written with GTest / Catch. It is not good if your feature is so interconnected
> that it prevents other people from writing their own unit gests. It should be possible
> to use your code in future without extensively mocking your classes.
> A bad example that required some extensive effort would be the storage engine API.

### Documentation

> All new Features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the changelog so that
> developers and users have a concise overview. 

- [ ] Added a *Changelog Entry* (referencing the corresponding public or internal issue number)
- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *http API* 
- [ ] Added *Swagger examples* for the http API  

### CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

File: arangod/Agency/Agent.cpp [hunk 1]
```diff
@@ -797,14 +848,80 @@ void Agent::advanceCommitIndex() {
...
+  auto tp = steady_clock::now() +
+    duration_cast<milliseconds>(duration<double>(timeout));
+  auto res = _promises.try_emplace(tp, futures::Promise<query_t>());
+
+  if (!res.second) {
+    THROW_ARANGO_EXCEPTION_MESSAGE(
+      TRI_ERROR_INTERNAL, "Failed to add promise for polling");
+  }
+  if (_lowestPromise > index) {
+    _lowestPromise = index;
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp:922]: Here is a chance that we set `_lowestPromise` to 0, if `index` was 0. Is this intended?

File: arangod/Agency/Agent.cpp [hunk 2]
```diff
@@ -425,7 +459,11 @@ priv_rpc_ret_t Agent::recvAppendEntriesRPC(term_t term, std::string const& leade
           << " with term " << term;
       {
         WRITE_LOCKER(oLocker, _outputLock);
-        _commitIndex = std::max(_commitIndex, std::min(leaderCommitIndex, lastIndex));
+        auto const tmp = std::max(_commitIndex, std::min(leaderCommitIndex, lastIndex));
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp:462]: `index_t` is only mildly longer than `auto`, therefore I would prefer it.
```suggestion
        index_t const tmp = std::max(_commitIndex, std::min(leaderCommitIndex, lastIndex));
```

File: arangod/Agency/Agent.cpp [hunk 3]
```diff
@@ -450,7 +488,11 @@ priv_rpc_ret_t Agent::recvAppendEntriesRPC(term_t term, std::string const& leade
   {
     WRITE_LOCKER(oLocker, _outputLock);
     CONDITION_LOCKER(guard, _waitForCV);
-    _commitIndex = std::max(_commitIndex, std::min(leaderCommitIndex, lastIndex));
+    auto const tmp = std::max(_commitIndex, std::min(leaderCommitIndex, lastIndex));
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp:491]:
```suggestion
    index_t const tmp = std::max(_commitIndex, std::min(leaderCommitIndex, lastIndex));
```

File: arangod/Agency/Agent.cpp [hunk 4]
```diff
@@ -1298,6 +1415,54 @@ read_ret_t Agent::read(query_t const& query) {
...
+void Agent::clearExpiredPolls() {
+  index_t commitIndex = 0;
+  {
+    READ_LOCKER(oLocker, _outputLock);
+    commitIndex = _commitIndex;
+  }
+  auto empty = std::make_shared<VPackBuilder>();
+  {
+    VPackObjectBuilder obj(empty.get());
+    empty->add("accepted", VPackValue(true));
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp:1429]: Unnecessary?

File: arangod/Agency/Agent.cpp [hunk 5]
```diff
@@ -1298,6 +1415,54 @@ read_ret_t Agent::read(query_t const& query) {
...
+
+/// Clear expired polls
+/// Wake up everybody with quertand delete with empty
+/// If qu is nullptr, we're resigning.
+void Agent::triggerPollsNoLock(query_t qu, SteadyTimePoint const& tp) {
+
+  if (qu == nullptr) { // We have resigned
+    qu = std::make_shared<VPackBuilder>();
+    VPackObjectBuilder qb(qu.get());
+    qu->add("accepted", VPackValue(false));
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp:1450]: OK, understood. If we have resigned, we say `accepted: false`. Do we also do a 503 and a redirect? Or do we happily send a `200` with `accepted: false`?

File: arangod/Agency/Agent.cpp [hunk 6]
```diff
@@ -1298,6 +1415,54 @@ read_ret_t Agent::read(query_t const& query) {
...
+    qu->add(VPackValue("result"));
+    VPackArrayBuilder arr(qu.get());
+  }
+
+  auto it = _promises.begin();
+  while (it != _promises.end()) {
+    if (it->first < tp) {
+      it->second.setValue(qu);
+      it = _promises.erase(it);
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp:1459]: Please excuse my ignorance: Who holds the allocation of the promise until something has acted upon the newly set value?

File: arangod/Agency/Agent.cpp [hunk 7]
```diff
@@ -1298,6 +1415,54 @@ read_ret_t Agent::read(query_t const& query) {
...
+  if (qu == nullptr) { // We have resigned
+    qu = std::make_shared<VPackBuilder>();
+    VPackObjectBuilder qb(qu.get());
+    qu->add("accepted", VPackValue(false));
+    qu->add(VPackValue("result"));
+    VPackArrayBuilder arr(qu.get());
+  }
+
+  auto it = _promises.begin();
+  while (it != _promises.end()) {
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp:1456]: The common case will be that all promises are woken up. We could optimize for this by not erasing the promises one by one but by erasing the whole map in the end. Also in the other case we could erase a range, which might be more efficient in some implementations...

File: arangod/Agency/Agent.cpp [hunk 8]
```diff
@@ -1354,6 +1522,9 @@ void Agent::run() {
       // Check whether we can advance _commitIndex
       advanceCommitIndex();
 
+      // Clear expired long polls
+      clearExpiredPolls();
```
Comment (neunhoef) [inline arangod/Agency/Agent.cpp:1526]: Why is this needed twice? See line 1498.

File: arangod/Agency/Agent.h [hunk 9]
```diff
@@ -150,6 +154,18 @@ class Agent final : public arangodb::Thread, public AgentInterface {
...
  private:
 
+  void logsForTrigger();
+
+  /// @brief clear expired polls registered by Agent::poll
+  ///        if qu is nullptr, we're resigning.
+  ///        Caller must have _promLock!
+  void triggerPollsNoLock(
+    query_t qu = nullptr,
+    SteadyTimePoint const& tp = std::chrono::steady_clock::now() + std::chrono::seconds(60));
```
Comment (neunhoef) [inline arangod/Agency/Agent.h:164]: What is the reason behind this default value? Do we need default values here?

File: arangod/Agency/Agent.h [hunk 10]
```diff
@@ -508,6 +524,11 @@ class Agent final : public arangodb::Thread, public AgentInterface {
   // lock for _ongoingTrxs
   arangodb::Mutex _trxsLock;
 
+  // @brief promises for poll interface and the guard (mvm-style
+  std::mutex _promLock;
+  index_t _lowestPromise;
+  std::map<SteadyTimePoint, futures::Promise<query_t>> _promises;
```
Comment (neunhoef) [inline arangod/Agency/Agent.h:530]: Please add a comment saying that the key is the expiration time. Maybe explain memory lifetime of the Promise and when it is deleted from this map.

File: arangod/Agency/RestAgencyHandler.cpp [hunk 11]
```diff
@@ -150,7 +150,92 @@ RestStatus RestAgencyHandler::handleTransient() {
...
+RestStatus RestAgencyHandler::pollIndex(
+  index_t const& start, double const& timeout) {
+  return waitForFuture(
+    _agent->poll(start, timeout)
+    .thenValue([this, start](std::shared_ptr<VPackBuilder> rb) {
+      VPackSlice res = rb->slice();
+      if (res.hasKey("accepted")) {
+        TRI_ASSERT(res.hasKey("result"));
+        VPackBuffer<uint8_t> payload;
+        VPackBuilder builder(payload);
```
Comment (neunhoef) [inline arangod/Agency/RestAgencyHandler.cpp:168]: In case of a full readDB we now copy the readDB a third time into `payload`.

File: arangod/Agency/RestAgencyHandler.cpp [hunk 12]
```diff
@@ -150,7 +150,92 @@ RestStatus RestAgencyHandler::handleTransient() {
...
+
+  // GET only
+  if (_request->requestType() != rest::RequestType::GET) {
+    return reportMethodNotAllowed();
+  }
+
+  // WARNING ////////////////////////////////////////////////////
+  // Leader only
+  if (!_agent->leading()) {  // Redirect to leader
+    if (_agent->leaderID() == NO_LEADER) {
```
Comment (neunhoef) [inline arangod/Agency/RestAgencyHandler.cpp:219]: I vaguely remember a discussion we had as to whether we should allow followers to respond to this request. Did we not say that a follower could respond but should say it is a follower? I forgot the details.

File: arangod/Agency/RestAgencyHandler.h [hunk 13]
```diff
@@ -25,10 +25,38 @@
 #define ARANGOD_REST_HANDLER_REST_AGENCY_HANDLER_H 1
 
 #include "Agency/Agent.h"
+#include "Futures/Future.h"
+#include "Futures/Unit.h"
 #include "RestHandler/RestVocbaseBaseHandler.h"
 
 namespace arangodb {
 
+// String to number conversions:
```
Comment (neunhoef) [inline arangod/Agency/RestAgencyHandler.h:34]: Do we not have all these in `lib/Basics/StringUtils.*`?

File: arangod/Cluster/AgencyCache.cpp [hunk 14]
```diff
@@ -0,0 +1,53 @@
+////////////////////////////////////////////////////////////////////////////////
```
Comment (neunhoef) [inline arangod/Cluster/AgencyCache.cpp:1]: Do we want to have this pseudo code here in this PR? Would be OK for me, but is it intentional or an oversight?

File: arangod/Cluster/AgencyCache.h [hunk 15]
```diff
@@ -0,0 +1,61 @@
+////////////////////////////////////////////////////////////////////////////////
```
Comment (neunhoef) [inline arangod/Cluster/AgencyCache.h:1]: See above, include?

File: tests/js/client/agency/agency-test.js [hunk 16]
```diff
@@ -377,6 +377,140 @@ function agencyTestSuite () {
...
+////////////////////////////////////////////////////////////////////////////////
+
+    testPoll : function () {
+      var cur = accessAgency("write",[[{"/": {"op":"delete"}}]]).
+          bodyParsed.results[0];
+      var ret, result, log, ci, job, wr;
+
+      ret = request({url: agencyLeader + "/_api/agency/poll",
+                         method: "GET", followRedirect: true});
+      if (ret.statusCode === 202) {
```
Comment (neunhoef) [inline tests/js/client/agency/agency-test.js:389]: Do we answer with 202?

File: arangod/Agency/RestAgencyHandler.cpp [hunk 17]
```diff
@@ -150,7 +150,92 @@ RestStatus RestAgencyHandler::handleTransient() {
...
+            builder.add(VPackValue("log"));
+            VPackArrayBuilder a(&builder);
+            for (size_t i = start - firstIndex.getNumber<uint64_t>(); i < logs.length(); i++) {
+              builder.add(logs[i]);
+            }
+          } else {
+            builder.add("readDB", slice.get("readDB"));
+          }
+        }
+        resetResponse(rest::ResponseCode::ACCEPTED);
```
Comment (neunhoef) [inline arangod/Agency/RestAgencyHandler.cpp:190]: Why accepted?

File: tests/js/client/agency/agency-test.js (hunk 16 above)
Comment (neunhoef) [reply tests/js/client/agency/agency-test.js:389]: See comment above, I would have thought 200 is more appropriate.

Comment (neunhoef) [reply tests/js/client/agency/agency-test.js:389]: Furthermore, let's assert on the status code. Otherwise, if it is wrong, we have a strange error message with an exception further down.

File: 3rdParty/rocksdb/6.2/util/channel.h [hunk 18]
```diff
@@ -0,0 +1,67 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
```
[the remark this answers is not in the document]
Comment (kvahed) [reply 3rdParty/rocksdb/6.2/util/channel.h:1]: left over did not compile with clang

File: arangod/Agency/Agent.cpp [hunk 19]
```diff
@@ -797,14 +848,80 @@ void Agent::advanceCommitIndex() {
...
+    } else if (index <= commitIndex) {   // deliver immediately all logs since index
+      logs = _state.get(index, commitIndex);
+    }
+  }
+
+  if (!logs.empty() || readDB != nullptr) {
+    auto builder = std::make_shared<VPackBuilder>();
+    {
+      VPackObjectBuilder e(builder.get());
+      builder->add("accepted", VPackValue(true));
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Agency/Agent.cpp:889]: it was originally planned as indication that we are leading. technically no longer necessary

File: arangod/Agency/Agent.cpp (hunk 1 above)
Comment (kvahed) [reply arangod/Agency/Agent.cpp:922]: no, if index == 0, you never get here

File: arangod/Agency/Agent.cpp (hunk 4 above)
Comment (kvahed) [reply arangod/Agency/Agent.cpp:1429]: no longer

File: arangod/Agency/Agent.cpp (hunk 5 above)
Comment (kvahed) [reply arangod/Agency/Agent.cpp:1450]: no this is no longer necessary, as we are not going to not answer, if not leading

File: arangod/Agency/Agent.cpp (hunk 7 above)
Comment (kvahed) [reply arangod/Agency/Agent.cpp:1456]: very good point. awesome suggestion.

File: arangod/Agency/Agent.cpp (hunk 8 above)
Comment (kvahed) [reply arangod/Agency/Agent.cpp:1526]: It is not

File: arangod/Agency/Agent.h (hunk 9 above)
Comment (kvahed) [reply arangod/Agency/Agent.h:164]: it's for resignation case

File: arangod/Agency/Agent.h (hunk 10 above)
Comment (kvahed) [reply arangod/Agency/Agent.h:531]: done.

File: arangod/Agency/RestAgencyHandler.cpp (hunk 17 above)
Comment (kvahed) [reply arangod/Agency/RestAgencyHandler.cpp:190]: resignation

File: arangod/Agency/RestAgencyHandler.cpp (hunk 12 above)
Comment (kvahed) [reply arangod/Agency/RestAgencyHandler.cpp:219]: yes and this is unrelated. 503.

File: arangod/Agency/RestAgencyHandler.h (hunk 13 above)
Comment (kvahed) [reply arangod/Agency/RestAgencyHandler.h:34]: not generalized so that one call like stp below gives you all. i am happy to carry these over to stringutils.

File: arangod/Cluster/AgencyCache.cpp (hunk 14 above)
Comment (kvahed) [reply arangod/Cluster/AgencyCache.cpp:1]: oversight

File: arangod/Cluster/AgencyCache.h (hunk 15 above)
Comment (kvahed) [reply arangod/Cluster/AgencyCache.h]: same same but different

File: tests/js/client/agency/agency-test.js (hunk 16 above)
Comment (kvahed) [reply tests/js/client/agency/agency-test.js:389]: 202 just passed on the RestHandlers's code.  adjusted

File: arangod/Agency/Agent.cpp [hunk 20]
```diff
@@ -375,6 +375,39 @@ void Agent::reportFailed(std::string const& slaveId, size_t toLog, bool sent) {
...
+  // Every resthandler takes, what it needs.
+  // Delete all promises.
+  // Reset _lowestPromise.
+  std::lock_guard lck(_promLock);
+  auto builder = std::make_shared<VPackBuilder>();
+  {
+    VPackObjectBuilder e(builder.get());
+    auto const logs = _state.get(_lowestPromise, _commitIndex);
+
+    TRI_ASSERT(!logs.empty());
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Agency/Agent.cpp:391]: done

File: arangod/Agency/Agent.h [hunk 21]
```diff
@@ -43,7 +44,8 @@ struct TRI_vocbase_t;
 namespace arangodb {
 namespace consensus {
 
-class Agent final : public arangodb::Thread, public AgentInterface {
+class Agent final : public arangodb::Thread, public AgentInterface,
+                    public std::enable_shared_from_this<Agent> {
```
[the remark this answers is not in the document]
Comment (kvahed) [reply arangod/Agency/Agent.h:48]: true. the original introduction came from the discarded method for scheduling the promises for fulfilling

File: arangod/Agency/RestAgencyHandler.h (hunk 13 above)
Comment (kvahed) [reply arangod/Agency/RestAgencyHandler.h:34]: that's what i am doing?

Comment (kvahed) [reply arangod/Agency/RestAgencyHandler.h:34]: nice i got it.
